Shopping Cart: Add cart-view page

danielbeardsley opened the pull request:
Create most of the cart view page according to the figma design.

This started as copying the CartDrawer (re-copied this morning) and modifying it as needed to match the figma. I also extracted a few bits that weren't likely to change.

I've opted in many cases to copy the functionality because keeping one component that can work in both places would require compromising on styling a fair amount, or increase complexity (adding a page and drawer mode).

### Features

* Cart contents
* Summary / totals
* Checkout button
* Cross sell (borrowed component from Drawer)
* Cart Alerts

This is missing some of the sections from the design and I imagine those will come as future pulls but I want to get this in the code base now as the `CartDrawer` continues to evolve.

Connect iFixit/ifixit#48798

danielbeardsley committed:
Cart View: initial commit

Just the boilerplate with no server-side props except the default layout
stuff to support the header / footer.

danielbeardsley committed:
Cart: add basic structure for line items

danielbeardsley committed:
Cart: copy cart drawer implementation

We're going to re-use the useCart() hooks for state and mutation, but
the UI will likely be different. Probably not all that different though,
so let's try starting with a copy of the cart drawer code.

Further commits will remove un-needed bits and fix the relative paths
and such. This commit is a direct copy/pasta.

danielbeardsley committed:
Shopping Cart: make it compiles

In the previous commit, we copied the cart-drawer whole-sale. Here
we make it compile in it's new place (and dis-robe it of the Drawer)
trappings.

It's not pretty yet.

danielbeardsley committed:
Cart: put up scaffolding

Getting the layout roughed in.

danielbeardsley committed:
Cart: layout tweaks

danielbeardsley committed:
Cart: extract a function

danielbeardsley committed:
Cart: extract a component

danielbeardsley committed:
Cart: extract CartAlerts component

danielbeardsley committed:
Cart: left-align the title

danielbeardsley committed:
squash with extract a function

danielbeardsley committed:
vscode: use typescript version from the code base

Instead of whatever version comes with vscode. This addressed a constant
problem with tsserver saying "This expression produces a union type
that's too complex to represent" or something.

danielbeardsley committed:
Cart: style up the totals pane

danielbeardsley committed:
Cart: remove Total section from left side

Now that we have it on the right side, we don't need it under the left.

danielbeardsley committed:
Cart: refactor wrappers around CartLineItem

Mostly with the intent of making a discounted price not result in much
taller line item.

Viewing this while ignoring whitespace makes more sense.

danielbeardsley committed:
Cart: tune layout to deal with with AnimatedList

alignItems="center" on the parent of an AnimatedList result in a 0-width list.
So, we need to change a few things to get them the proper width.

danielbeardsley committed:
Cart: make image size match figma

danielbeardsley committed:
Cart: layout tweaks to match figma

danielbeardsley committed:
Cart: put checkout button under Summary on the right

danielbeardsley committed:
Cart Drawer: move margin / id box out of component

The other usage (/Cart/view) doesn't want the same outer margin nor the
id.

danielbeardsley committed:
Cart: style checkout button to match figma

danielbeardsley committed:
Cart: Correct mobile layout / flow

Match the figma and create a reasonable in-between state.

danielbeardsley committed:
Cart: remove unused imports

danielbeardsley committed:
Cart: layout "EmptyState" for the new whole-page view

Hide the items-count badge when it's 0 cause it's not obvious what it
represents.

danielbeardsley committed:
Cart: drop unsupported debug value

sterlinghirsh commented on packages/ui/cart/view/CartLineItem.tsx:
```diff
@@ -0,0 +1,188 @@
+         <CartLineItemImage src={lineItem.imageSrc} alt={lineItem.name} />
+         <VStack align="flex-start" spacing="0" flexGrow="1">
+            <Link
+               href={`${appContext.ifixitOrigin}/Store/Product/${lineItem.itemcode}`}
```
What is this link? Should this be /products/{handle}? Or is there a url property on the lineItem?

danielbeardsley replied:
Good catch. I'm not sure, though it's the same as the current cart-drawer:

https://github.com/iFixit/react-commerce/blob/7726466185b364acee4b92f2e7a9124ade6e724d/packages/ui/cart/drawer/CartLineItem.tsx#L95

Looks like a redirecter: https://www.ifixit.com/Store/Product/IF145-307-4

Though... I can't yet figure out where it's from.

danielbeardsley replied:
Definitely php code because I see varnish headers in the response.

danielbeardsley replied:
I logged the route and got here:
![image](https://github.com/iFixit/react-commerce/assets/26855/576294ce-dede-445d-936f-56bb58c23447)

Which ends up here for the redirect:
https://github.com/iFixit/ifixit/blob/72d86ea35f0b850df0848070df8eb5980f613d00/Cart/UI/parts_store.php#L120

Crazy.

dhmacs replied:
I think this was here before we had the product pages in next.js and has never been updated. It think we can safely refactor it now 👍

sterlinghirsh commented:
CR :zap: dev_block :zap: on verifying the product link. Also I want to make sure that our header link to the cart isn't a next link that will start trying to serve this. Otherwise this looks like a good start on the cart.

dhmacs commented on packages/cart-sdk/hooks/use-cart.ts:
```diff
@@ -35,6 +36,26 @@ export function useCart() {
    return query;
 }
 
+export function extractCrossSells(cart: Cart | null | undefined) {
+   return useMemo(
```
I noticed that this function contains a hook, therefore [by React convetions it should be a hook itselves](https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-from-react-functions).

I'd take this opportunity to suggest to refactor like this (taking advantage of context, so where we use this we don't have to use another hook to get the cart):

```suggestion
export function useCartCrossSells() {
   const cart = useCart();
   return useMemo(
```

danielbeardsley replied:
Ok. I'm always a little leery about the performance effects of adding additional calls to contexts when the values we wan't already in scope. That said, I appreciate the encapsulation.

I'll go for it.

dhmacs reviewed (commented):
Great job @danielbeardsley 👏 I left a couple of suggestions. 

What do you think about moving this work under the app router? Seems like the work in #1950 is going to get merged soon, and this looks like a good page to get started with the app router

danielbeardsley commented on .vscode/settings.json:
```diff
@@ -1 +1,3 @@
-{}
+{
+   "typescript.tsdk": "node_modules/typescript/lib"
+}
```
This tells vscode to use the typescript (for tsserver) version from the repo instead of the one that's bundled in vscode. 

See: becd0b5e8856381703f48225e92e0992660d8040

danielbeardsley committed:
Cross sells: turn function into a hook

This doesn't change it's behavior, it mostly removes an argument and
changes its name.

sterlinghirsh commented:
This has been paused because some other projects became higher priority. I agree that we should move this under the App router whenever we pick this up.

danielbeardsley commented:
This can be re-opened (if desired) in this code's new home
